### PR TITLE
custom species name

### PR DESCRIPTION
**Pull request by MilonPL**
## About the PR

title

adds our own `emberfall_profile` table to the db which stores the custom species name in relation to profile id
havent tested it with postgres but i'm sure it will be fine
the rest is simple, adds a nullable string to the HumanoidApperanceComponent and we use that if its not null or otherwise fall back to the prototype name
needs to be autonetworked for the examine stuff or you get a delay etc etc
set in the station spawning system initially

## Why / Balance

\>see a cat
\>shift click
\>it is a young rodentia

## Media

image

(its lowercased like the rest now but i didnt bother taking another screenshot)
image

### CODE REVIEW

File: Content.Shared/Preferences/HumanoidCharacterProfile.cs [hunk 1]
```diff
@@ -77,6 +77,12 @@ public sealed partial class HumanoidCharacterProfile : ICharacterProfile
         [DataField]
         public ProtoId<SpeciesPrototype> Species { get; set; } = SharedHumanoidAppearanceSystem.DefaultSpecies;
 
+        /// <summary>
+        /// Emberfall: Custom species name.
+        /// </summary>
+        [DataField]
+        public string CustomSpeciesName { get; set; } = string.Empty;
+
```
Comment (dffdff2423) [inline Content.Shared/Preferences/HumanoidCharacterProfile.cs:85]: If you intend to make a ton of additions to `HumanoidCharacterProfile` I would recommend creating a separate class for it (or even possibly extending `HumanoidCharacterProfile`)

File: Content.Client/Lobby/UI/HumanoidProfileEditor.xaml.cs [hunk 2]
```diff
@@ -218,8 +218,24 @@ public HumanoidProfileEditor(
...
                 UpdateHairPickers();
                 OnSkinColorOnValueChanged();
+                UpdateCustomSpeciesNameEdit(); // Emberfall
             };
 
+            #region Emberfall Custom Species
+
+            CustomSpeciesNameEdit.OnTextChanged += args =>
+            {
+                var text = new string(args.Text.Take(16).Where(c => char.IsLetter(c)).ToArray());
```
Comment (dffdff2423) [inline Content.Client/Lobby/UI/HumanoidProfileEditor.xaml.cs]: I hate this. Also I think the text box should only permit letters automatically? Slicing it like I do [here](https://github.com/cosmatic-drift-14/cosmatic-drift/blob/8bc36f840846441d57ea129832e40f082a75549b/Content.Shared/_CD/Records/PlayerProvidedCharacterRecords.cs#L184-L191) (which can be turned into a oneliner) should be much faster than linq but perf really does not matter for this. (I think this also makes unnecessary allocations, but I am not enough of a C# expert to know for sure)

File: Content.Server.Database/EmberfallModel.cs [hunk 3]
```diff
@@ -8,11 +8,8 @@ public class EmberfallProfile
...
         [Column("emberfallprofile_id")]
         public int Id { get; set; }
-
-        [Column("profile_id")]
-
         public int ProfileId { get; set; }
-        public Profile Profile { get; set; } = null!;
-        public string CustomSpeciesName { get; set; } = string.Empty;
+        public virtual Profile Profile { get; set; } = null!;
+        public string? CustomSpeciesName { get; set; } = string.Empty;
```
Comment (dffdff2423) [inline Content.Server.Database/EmberfallModel.cs]: probably want this to be null instead of string.Empty?

Comment (MilonPL) [reply Content.Server.Database/EmberfallModel.cs]: riiight

File: Content.Shared/Preferences/HumanoidCharacterProfile.cs (hunk 1 above)
Comment (MilonPL) [reply Content.Shared/Preferences/HumanoidCharacterProfile.cs:85]: probably not going to be an issue in the near future so i'll cross that bridge when i get to it

File: Content.Client/Lobby/UI/HumanoidProfileEditor.xaml.cs (hunk 2 above)
Comment (MilonPL) [reply Content.Client/Lobby/UI/HumanoidProfileEditor.xaml.cs]: you're just a LINQ hater smh
yes idk why i went with this method where it should've just been `var newText = string.Concat(CustomSpeciesNameEdit.Text.Where(char.IsLetter));` really

**Comment by MilonPL** [comment]
what's the worst that could happen

**Pull request merged**